README.md

```diff
@@ -160,6 +160,131 @@ Returning:
 </div>
 ```
 
+#### Slots (experimental)
+
+_Slots are currently under development as a successor to Content Areas. The Slot APIs should be considered unfinished and subject to breaking changes in non-major releases of ViewComponent._
+
+Slots enable multiple blocks of content to be passed to a single ViewComponent.
+
+Slots exist in two forms: normal slots and collection slots.
+
+Normal slots can be rendered once per component. They expose an accessor with the name of the slot that returns an instance of `BoxComponent::Header`, etc.
+
+Collection slots can be rendered multiple times. They expose an accessor with the pluralized name of the slot (`#rows`), which is an Array of `BoxComponent::Row` instances.
+
+To learn more about the design of the Slots API, see https://github.com/github/view_component/pull/348.
+
+##### Defining Slots
+
+Slots are defined by the `with_slot` macro:
+
+`with_slot :header`
+
+To define a collection slot, add `collection: true`:
+
+`with_slot :row, collection: true`
+
+To define a slot with a custom class, pass `class_name`:
+
+`with_slot :body, class_name: 'BodySlot`
+
+Slot classes should be subclasses of `ViewComponent::Slot`.
+
+##### Example ViewComponent with Slots
+
+`# box_component.rb`
+```ruby
+class BoxComponent < ViewComponent::Base
+  with_slot :body, :footer
+  with_slot :header, class_name: "Header"
+  with_slot :row, collection: true, class_name: "Row"
+
+  class Header < ViewComponent::Slot
```

Comment:
I wonder if it would be possible to have slots still inherit from `ViewComponent::Base`?

If we could do that, I think it'd be one less thing to remember and extracting full components is easier since you no longer have to change what it inherits from.

Reply:
I could see that being possible, but my concern with that approach would be that it might imply that components could be used as Slots, when a Slot is just a simple accessor object. I do wonder if there's a simpler way to express the Slot interface though, as all we really care about is that the object has a `content` accessor.

Reply:
I can't help scratch the itch for mr that slots could themselves be ViewComponents. I've mentioned this before and I don't yet have the concrete uses cases (though i have my suspicions) so I'm prepared to let this go for a future release.

But here is my thinking. If you stripped out some of the magic around class_name and generating classes we're providing in `with_slot` you could imagine that you had to use slots like this:

```erb
<% component.slot(:header, HeaderSlot.new(class_names: "my-class-name")) do %>
    This is my header!
  <% end %>
```
the requirement on HeaderSlot being that it provide a `content` property to put the captured block in. 

Under that light the `component.slot` is looking an awful lot like a `render` call on a ViewComponent.

Now imagine the guts of the `slot` method was something like this:

```ruby
    slot = slots[slot_name]

    if args.is_a?(Hash)
      # The class name of the Slot, such as Header
      slot_class = self.class.const_get(slot[:class_name])

      # Instantiate Slot class, accommodating Slots that don't accept arguments
      slot_instance = args.present? ? slot_class.new(args) : slot_class.new
    else
      # args were really a class
      slot_instance = args
    end

    if slot_instance.responds_to(:render_in)
      # RenderInSlotWrapper exposes a `content` property that is the result of `render_in`
      # I'd need to think it through more but I think we need some sort of wrapper that 'captures' the entirety of the rendered ViewComponent
      slot_instance = RenderInSlotWrapper.new(slot_instance).render_in(view_context, &block) 
    else
      # Capture block and assign to slot_instance#content
      slot_instance.content = view_context.capture(&block) if block_given?
    end
```

My spidey senses says there is something power here.

Reply:
@jonspalmer I disagree with the following:

```erb
<% component.slot(:header, HeaderSlot.new(class_names: "my-class-name")) do %>
  This is my header!
<% end %>
```

If the `HeaderSlot` is publicly accessible like that, I wouldn't view it as a slot anymore. Slots are internal to the component and shouldn't be usable outside of the component, imo. You can of course put a component in a slot, which would be a very common use case of component composition:

```erb
<%= render(MediaObjectComponent.new) do |media_object| %>
  <% media_object.slot(:left) do %>
    <%= render(AvatarComponent.new(image: comment.image)) %>
  <% end %>
  <% media_object.slot(:right) do %>
    <%= comment.body %>
  <% end %>
<% end %>
```

Reply:
I agree with the pattern of using a component in a slot. I believe that will be powerful enough for many uses cases.

However, I don't really see the 'internal' piece of a component quite the same way. I see Slots as defining an interface between the component and the slots it uses. Its an Interface not a concrete implementation. As such a ViewComponent leveraging slots comes with so slot classes. IMO it would be powerful allow those classes to be extended. Its let about using those Slot classes outside the component its that I can extend them (subclassing most likely) to be used in the component.

For example consider the Row example from the docs 

```ruby
class Row < ViewComponent::Slot
    def initialize(theme: :gray)
      @theme = theme
    end

    def theme_class_name
      case @theme
      when :gray
        "Box-row--gray"
      when :hover_gray
        "Box-row--hover-gray"
      when :yellow
        "Box-row--yellow"
      when :blue
        "Box-row--blue"
      when :hover_blue
        "Box-row--hover-blue"
      else
        "Box-row--gray"
      end
    end
  end
``` 

The interface that the component relies on is that Row have a content getter/setter and it has a getter for `theme_class_name`.

No suppose I want to use this component but add some other classes to the header based on logic special to my use case. I'd do this:

```ruby
class MySpecialRow < Row

  def initialize(theme: :gray, special: :foo)
    super(theme)
    @special = special
  end

  def theme_class_name
     special_classes = some_other_logic(special)
     "#{super} #{special_classes}
  end
end
```

and then in the component usage

```erb
  <% component.slot(:row, MySpecialRow.new(theme: :hover_blue, special: :make_some_magic) do %>
    Special row
  <% end %>
```

Reply:
Hm, couldn't you use composition instead? I'm just thinking composition is generally a better pattern in the view layer than overriding/extending. It's the only way you can build components in React, and it works pretty well there. But I admit I have a hard time seeing all the possibilities of 3rd party components and how people would want to extend them.

I guess to phrase it differently, let's say there's a 3rd party library with bootstrap components built with view components. Would you really want to modify them, or would you rather wrap them (composition) with your own components that hold the specialized behavior?